lib/Target/Btor/BtorToBtorIRTranslation.cpp

```diff
@@ -0,0 +1,781 @@
+        case BTOR2_TAG_nand:
+        case BTOR2_TAG_neg:
+        case BTOR2_TAG_neq:
+        case BTOR2_TAG_nor:
+        case BTOR2_TAG_not:
+        case BTOR2_TAG_one:
+        case BTOR2_TAG_ones:
+        case BTOR2_TAG_or:
+        case BTOR2_TAG_output:
+        case BTOR2_TAG_redand:
+        case BTOR2_TAG_redor:
+        case BTOR2_TAG_redxor:
+        case BTOR2_TAG_sdiv:
+        case BTOR2_TAG_sext:
+        case BTOR2_TAG_sgt:
+        case BTOR2_TAG_sgte:
+        case BTOR2_TAG_iff:
+        case BTOR2_TAG_slice:
+        case BTOR2_TAG_sll:
+        case BTOR2_TAG_slt:
+        case BTOR2_TAG_slte:
+        case BTOR2_TAG_sra:
+        case BTOR2_TAG_srem:
+        case BTOR2_TAG_srl:
+        case BTOR2_TAG_sub:
+        case BTOR2_TAG_udiv:
+        case BTOR2_TAG_uext:
+        case BTOR2_TAG_ugt:
+        case BTOR2_TAG_ugte:
+        case BTOR2_TAG_ult:
+        case BTOR2_TAG_ulte:
+        case BTOR2_TAG_urem:
+        case BTOR2_TAG_xnor:
+        case BTOR2_TAG_xor:
+        case BTOR2_TAG_zero:
+        case BTOR2_TAG_read:
+        case BTOR2_TAG_write:
+        case BTOR2_TAG_fair:
+        case BTOR2_TAG_justice:
+        case BTOR2_TAG_rol:
+        case BTOR2_TAG_ror:
+        case BTOR2_TAG_saddo:
+        case BTOR2_TAG_sdivo:
+        case BTOR2_TAG_smod:
+        case BTOR2_TAG_smulo:
+        case BTOR2_TAG_ssubo:
+        case BTOR2_TAG_uaddo:
+        case BTOR2_TAG_umulo:
+        case BTOR2_TAG_usubo:
+        default:
+        break;
+    }
+}
+
+static void parse_model () {
+    assert (model_file);
+    model = btor2parser_new ();
+    if (!btor2parser_read_lines (model, model_file))
+    std::cerr << "parse error at: " << btor2parser_error (model) << "\n";
+    num_format_lines = btor2parser_max_id (model);
+    inits.resize (num_format_lines, nullptr);
+    nexts.resize (num_format_lines, nullptr);
+    Btor2LineIterator it = btor2parser_iter_init (model);
+    Btor2Line *line;
+    while ((line = btor2parser_iter_next (&it))) parse_model_line (line);
+
+    for (size_t i = 0; i < states.size (); i++)
+    {
+    Btor2Line *state = states[i];
+    if (!nexts[state->id])
+    {
+        std::cerr << "state " << state->id << " without next function\n";
+    }
+    }
+}
+
+void filterInits() {
+    std::vector<Btor2Line *> filteredInits;
+    for( auto it = inits.begin(); it != inits.end(); ++it ) {
+        if( *it ) {
+            filteredInits.push_back( *it );
+        }
+    }
+    inits.clear();
+    inits = filteredInits;
+}
+
+void filterNexts() {
+    std::vector<Btor2Line *> filteredNexts;
+    for( auto it = nexts.begin(); it != nexts.end(); ++it ) {
+        if( *it ) {
+            filteredNexts.push_back( *it );
+        }
+    }
+    nexts.clear();
+    nexts = filteredNexts;
+}
+
+
+Operation * createMLIR( const Btor2Line * line, 
+                    OpBuilder builder,
+                    const std::map<int64_t, Value> cache,
+                    const int64_t * kids  ) {
```

Comment:
pointer to kids? not a vector or array wrapper?

Reply:
Yes. Btor2Line has a pointer to args. Instead of keeping a vector/small vector of kids in our algorithm, we can pass all the arguments for our specific for operation creation